Marketing -> Traffic settings: Fix SEO & Analytics Settings for Free plans

allilevine opened the pull request:
#### Changes proposed in this Pull Request

Show the correct banners on Tools > Marketing > Traffic > Search Engine Optimization and Google Analytics:
- Use `hasActiveSiteFeature()` to check if we should show a banner / nudge or the setting
- Use `isAtomic` to show the correct nudge or form

#### Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* On a free Atomic site, open Tools > Marketing > Traffic settings.
* There should be upsell nudges under Search engine optimization and Google Analytics.
* Clicking the upsell nudges goes to plans, not checkout.
* Repeat for Personal and Premium plans on Atomic sites. Google Analytics should be available with no nudge on Premium Atomic.
* Regression: 
    * Regular/paid Atomic: SEO and Analytics should be available, no nudge.
    * Simple: The nudges should be the same as for free, Personal, and Premium Atomic.
    * VIP (see: PCYsg-zan-p2): SEO has a nudge with no plan, Google Analytics is available.
    * Jetpack: SEO should be available on all plans, Analytics is not available on free Jetpack and there should be a Jetpack checkout nudge.

Simple | Free Atomic | Regular Atomic
------- | ------------ | ---------------
<img width="1056" alt="Screen Shot 2021-06-01 at 5 04 32 PM" src="https://user-images.githubusercontent.com/1689238/120390393-a671fc80-c2fb-11eb-8235-39decb84b022.png"> | <img width="1061" alt="Screen Shot 2021-06-01 at 5 05 11 PM" src="https://user-images.githubusercontent.com/1689238/120390420-af62ce00-c2fb-11eb-862f-2b2131dcf83f.png"> | <img width="1059" alt="Screen Shot 2021-06-01 at 5 05 32 PM" src="https://user-images.githubusercontent.com/1689238/120390441-b4278200-c2fb-11eb-84e9-ecc2e2913b44.png">
<img width="1047" alt="Screen Shot 2021-06-01 at 5 04 41 PM" src="https://user-images.githubusercontent.com/1689238/120390518-cc979c80-c2fb-11eb-865d-bd4e603d19bb.png"> | <img width="1053" alt="Screen Shot 2021-06-01 at 5 05 19 PM" src="https://user-images.githubusercontent.com/1689238/120390526-ce616000-c2fb-11eb-87a8-d6958425ae20.png"> | <img width="1050" alt="Screen Shot 2021-06-01 at 5 05 41 PM" src="https://user-images.githubusercontent.com/1689238/120390531-d0c3ba00-c2fb-11eb-9764-612d604a2e3f.png">

<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->
[
Follow-up needed](https://github.com/Automattic/wp-calypso/issues/54555) to fix the upsell nudge link plan slugs for free Atomic (should be business-bundle and value_bundle).

Fixes #51828

allilevine committed:
Add isFreeAtomic site helper
Use the new helper to determine which banners and settings to show on
Marketing > Traffic.

allilevine committed:
Look for plan, rather than first similar plan
Previously we were checking for the closest type of plan available.
However, the closest plans for free Atomic are Jetpack plans.

allilevine committed:
Refactor isFreeAtomicSite helper

allilevine committed:
Revert looking for specific plan

allilevine commented on client/my-sites/marketing/traffic/index.js:
```diff
@@ -108,7 +109,8 @@ const connectComponent = connect( ( state ) => {
 	const isJetpack = isJetpackSite( state, siteId );
-	const isJetpackAdmin = isJetpack && isAdmin;
+	const isFreeAtomic = isFreeAtomicSite( state, siteId );
+	const isJetpackAdmin = isJetpack && isAdmin && ! isFreeAtomic;
```
I used `isFreeAtomic` rather than `isAtomic` here since these settings are currently available to Atomic sites.

allilevine commented on client/my-sites/site-settings/analytics/form-google-analytics.jsx:
```diff
@@ -89,7 +90,7 @@ export const GoogleAnalyticsForm = ( props ) => {
 		setDisplayForm,
 	};
-	if ( props.siteIsJetpack ) {
+	if ( props.siteIsJetpack && ! props.siteIsFreeAtomic ) {
```
I used `isFreeAtomic` rather than `isAtomic` here since the Jetpack version of this form is currently available to Atomic sites.

allilevine commented on client/my-sites/site-settings/seo-settings/form.jsx:
```diff
@@ -332,7 +336,7 @@ export class SeoForm extends React.Component {
 				{ siteId && <QueryJetpackPlugins siteIds={ [ siteId ] } /> }
 				{ siteIsJetpack && <QueryJetpackModules siteId={ siteId } /> }
-				{ ( isSitePrivate || isSiteHidden ) && hasSiteSeoFeature( selectedSite ) && (
+				{ ( isSitePrivate || isSiteHidden ) && showAdvancedSeo && (
```
`showAdvancedSeo` includes a check for `hasSiteSeoFeature( selectedSite )`, while making sure we don't show this warning on `jetpack_free` sites. We could leave this as-is though, if there are other types of sites that should show it.

allilevine commented on client/my-sites/site-settings/seo-settings/form.jsx:
```diff
@@ -482,13 +486,17 @@ const mapStateToProps = ( state ) => {
-		( ! siteIsJetpack || get( getJetpackModules( state, siteId ), 'seo-tools.available', false ) );
+		( ! siteIsJetpack ||
+			get( getJetpackModules( state, siteId ), 'seo-tools.available', false ) ) &&
+		! siteIsFreeAtomic;
```
Free Atomic sites will return true without this check.

allilevine commented on client/my-sites/site-settings/seo-settings/help.jsx:
```diff
@@ -25,16 +26,18 @@ const SeoSettingsHelpCard = ( {
+	const seoHelpLink =
+		siteIsJetpack && ! siteIsFreeAtomic
+			? 'https://jetpack.com/support/seo-tools/'
+			: 'https://wpbizseo.wordpress.com/';
```
Checking for `isFreeAtomic` here as well, to keep Atomic access to Jetpack features.

allilevine commented:
I'm running into an issue with the upsell nudge links to plans. We look for `findFirstSimilarPlanKey()`, rather than for the specific plan, so we're getting Jetpack plans like `&plan=jetpack_business` rather than `&plan=business-bundle`:

https://github.com/Automattic/wp-calypso/blob/75e0cab9778cbf5b37ccf27fa1f84a610b8c452e/client/my-sites/site-settings/seo-settings/form.jsx#L324-L328

I tried changing to specific plans, but we offer different interval upgrades, so that breaks the tests.

As a result, the highlighting on the plans page is incorrect:
<img width="825" alt="Screen Shot 2021-06-01 at 5 35 09 PM" src="https://user-images.githubusercontent.com/1689238/120393321-b8559e80-c2ff-11eb-9b88-8c75c06e61d5.png">

That may be best fixed in a separate PR though.

retrofox commented on client/lib/site/utils.js:
```diff
@@ -176,3 +178,16 @@ export function hasSiteFeature( site, feature ) {
+ * @returns {boolean} True if site is free and Atomic
+ */
+export function isFreeAtomicSite( state, siteId ) {
+	const isAtomic = isAtomicSite( state, siteId );
```
Does it make sense to bail early when it isn't an Atomic site? Something like...
```suggestion
	const isAtomic = isAtomicSite( state, siteId );
	if ( ! isAtomic ) {
	  return false;
  }
```

retrofox commented:
Hi, there 👋 
I'm not 100% sure about adding this selector.

> Add isFreeAtomicSite helper and use the new helper to determine which banners and settings to show on
Marketing > Traffic.

I immediately start to think that we're going to add `isPremiumAtomicSite`, isFreeSimpleSite` and all combinations.
On the other hand, adding this selector reduces the amount of code that makes adding it sense.

👍 to add the selector, but I'd like to hear other opinions.

cpapazoglou commented:
I agree with @retrofox! While currently we don't allow Atomic sites to buy a Personal or Premium plan, let's think ahead.

I wonder if we should check for explicit plans rather than `siteIsJetpack && ! siteIsFreeAtomic `. 
We shouldn't really care whether the site is in Simple or Atomic. It seems that in the past we have used the infrastructure type as an easy way to enable / disable features. 

Wouldn't it be better to check whether current plan supports X feature? Something like this 
https://github.com/Automattic/wp-calypso/blob/b703449c2b079916da70356acd66ca823760cebf/client/state/sites/plans/selectors/has-feature.js#L19-L21

allilevine commented:
> I immediately start to think that we're going to add `isPremiumAtomicSite`, isFreeSimpleSite` and all combinations.
> I agree with @retrofox! While currently we don't allow Atomic sites to buy a Personal or Premium plan, let's think ahead.

@retrofox @cpapazoglou That's a good point, we might allow different Atomic plans in the future. 

> I wonder if we should check for explicit plans rather than `siteIsJetpack && ! siteIsFreeAtomic`.

The issue (as described here: pd2C1Z-Q-p2) is that the plan for free Atomic sites is `jetpack_free`. With all of these `isJetpack && ! isAtomic` and `isJetpack && ! isFreeAtomic` checks, we're adding hacks to change that behavior in the UI. I agree that it would be better to check for plans and features, but as far as I can tell we need to change the plan for that to work.

retrofox commented:
I agree we are dealing with these issues using basic helpers bringing duplicated code. Maybe we should start to consider adding smarter selectors/helpers considering the new site variations depending on:

* the platform (Simple, Jetpack, Atomic)
* the plan: Free, Personal, Business, eCommerce, etc
* the features

### Platform helpers / selectors

* `isSimpleSite()`
* `isJetpackSite()`
* `isAtomicSite()`

I've seen that sometimes we'd like to know when the site is exclusively a Jetpack site (not Atomic). We could create a new helper for this, or maybe add a parameter to the helper function:

```js
function isJetpackSite(  notAtomic = false ) {
  // Check whether the Jetpack is not Atomic site.
  if ( notAtomic && isAtomicSite() ) {
    return false;
  }
};
```

### Plans helpers / selectors

### Features helpers / selectors

* `hasMarketplaceAddOns()`

Is it a good moment to review/audit the current implementation and start coming up with a refactoring plan?

allilevine committed:
Refactor to rely on feature checks

mmtr reviewed (commented):
Nice, overall this works quite well for me.

Is it possible to change the plan slug used in the URL?

On Simple sites we use `business-bundle`, so the Business plan gets highlighted:

<img width="1269" alt="Screen Shot 2021-06-03 at 16 25 57" src="https://user-images.githubusercontent.com/1233880/120661588-b66b1700-c488-11eb-98ea-0fc89ac14f59.png">
<img width="853" alt="Screen Shot 2021-06-03 at 16 27 58" src="https://user-images.githubusercontent.com/1233880/120661610-b9fe9e00-c488-11eb-80dd-e4e86b298a2c.png">

However, on Free Atomic sites we use `jetpack_business` which is not available on WP.com so none of the plans gets highlighted:

<img width="1268" alt="Screen Shot 2021-06-03 at 16 26 29" src="https://user-images.githubusercontent.com/1233880/120661750-d569a900-c488-11eb-8048-cb309063e6d5.png">
<img width="833" alt="Screen Shot 2021-06-03 at 16 29 24" src="https://user-images.githubusercontent.com/1233880/120661810-e2869800-c488-11eb-9816-eb597f79ff3d.png">

mmtr commented:
Also noted a small regression on my Free Jetpack site.

On production, we directly link to the checkout page:

<img width="1263" alt="Screen Shot 2021-06-03 at 16 35 04" src="https://user-images.githubusercontent.com/1233880/120662972-e0710900-c489-11eb-9bcb-7ea5307a6f3f.png">

On this branch, we link to the plans page:
<img width="1249" alt="Screen Shot 2021-06-03 at 16 35 33" src="https://user-images.githubusercontent.com/1233880/120662996-e9fa7100-c489-11eb-9f4f-c4beff44b502.png">

retrofox commented:
> Also noted a small regression on my Free Jetpack site.

Is it an Atomic site? Because it may redirect the user to the plans page instead of the checkout one when it's an Atomic site.

mmtr commented:
> Is it an Atomic site? Because it may redirect the user to the plans page instead of the checkout one when it's an Atomic site.

No, it's a Jetpack site

allilevine committed:
Check for feature eligibility for Google Analytics form

allilevine committed:
Fix hasSiteSeoFeature tests

allilevine committed:
Change plan slug and tests

allilevine committed:
Fix plans page Premium feature highlight

allilevine committed:
Fix form-google-analytics tests

allilevine committed:
Revert allowing upgrade to Premium

allilevine commented:
> Is it possible to change the plan slug used in the URL?

@mmtr Thanks for taking a look! I'd written [a comment above](https://github.com/Automattic/wp-calypso/pull/53196#issuecomment-852461882) about why I was hesitant to make that change in this PR. I decided to try again since we shouldn't need to worry about plan intervals when we're starting with a free plan anyway. It seems to be working (with some test fixes), but the highlight for Premium / Google Analytics is still wrong:
<img width="1092" alt="Screen Shot 2021-06-03 at 5 58 50 PM" src="https://user-images.githubusercontent.com/1689238/120717328-5c248300-c495-11eb-91e5-98d5cb90340b.png">
The cause seems to be that we don't allow Atomic sites on a free plan / no plan to upgrade to Premium:
- We only add the `is-highlighted` class if `availableForPurchase` is `true` (as of https://github.com/Automattic/wp-calypso/pull/53146)
- `availableForPurchase` is only `true` if `canUpgradeToPlan()` returns `true`
- `canUpgradeToPlan()` doesn't allow Atomic sites without a plan to upgrade to Premium:
https://github.com/Automattic/wp-calypso/blob/e797df8452361ce7ead6ae1cdfec37378da0309c/client/state/selectors/test/can-upgrade-to-plan.js#L347-L351

Changing that seems like it should be handled in a separate PR. Let me know if I've missed anything there.

> Also noted a small regression on my Free Jetpack site.

Let me know if this is fixed now. I've made the logic for which banner to show more specific. It should look like this for free Jetpack:
<img width="1054" alt="Screen Shot 2021-06-03 at 6 09 50 PM" src="https://user-images.githubusercontent.com/1689238/120718196-e4efee80-c496-11eb-9726-8ff1c6791f85.png">

mmtr commented:
> The cause seems to be that we don't allow Atomic sites on a free plan / no plan to upgrade to Premium:

Yeah, this is fine to address on a follow-up. My comment was specifically for the SEO banner which requires the Business plan and it's available for Atomic sites, but missed to explicitly note it. I see it has been now fixed 🎉 

> I've made the logic for which banner to show more specific. It should look like this for free Jetpack:

Awesome, I can confirm it's been fixed now.

mmtr commented on client/lib/site/utils.js:
```diff
@@ -176,3 +177,17 @@ export function hasSiteFeature( site, feature ) {
+ * @returns {boolean} True if does
+ */
+export function isEligibleForSEOFeatures( site, state, siteId ) {
+	if ( site && site.plan && state ) {
```
Nitpick: Show we return false when this check doesn't pass?

allilevine replied:
Good point, changed here: https://github.com/Automattic/wp-calypso/pull/53196/commits/188710bebb54c56996e0266435a6b2dab5ae487f

mmtr commented on client/my-sites/marketing/traffic/index.js:
```diff
@@ -105,10 +106,12 @@ const SiteSettingsTraffic = ( {
 	const isJetpack = isJetpackSite( state, siteId );
-	const isJetpackAdmin = isJetpack && isAdmin;
+	const isEligible = isEligibleForSEOFeatures( site, state, siteId );
+	const isJetpackAdmin = isJetpack && isAdmin && isEligible;
```
`isJetpackAdmin` seems used for other features than SEO like Stats or the WP.me Shortlinks. Why `isEligibleForSEOFeatures` is controlling these features?

allilevine replied:
The issue I'm trying to solve here is that `isJetpackAdmin` returns `true` for a free / no plan Atomic site. There's no particular feature we're checking for, just Jetpack. But this check is used to determine whether to show the Stats and WP.me Shortlinks features. I guess it should really be checking for Analytics feature availability. Would having a `isEligibleForAnalyticsFeatures` helper or using `hasSiteAnalyticsFeature()` help clarify that?

mmtr replied:
> The issue I'm trying to solve here is that `isJetpackAdmin` returns `true` for a free / no plan Atomic site.

Why is this an issue? AFAIK both Stats or the WP.me Shortlinks are features available for free sites, aren't they?

allilevine replied:
Stats and WP.me Shortlinks are available to all Jetpack plans (including free) and Atomic sites. But they’re not available for Simple free sites, under Tools > Marketing > Traffic:

Simple | Atomic
------- | -------
<img width="1051" alt="Screen Shot 2021-06-07 at 2 10 31 PM" src="https://user-images.githubusercontent.com/1689238/121068476-529f6180-c79a-11eb-936a-d33784d17356.png"> | <img width="1055" alt="Screen Shot 2021-06-07 at 2 10 56 PM" src="https://user-images.githubusercontent.com/1689238/121068466-503d0780-c79a-11eb-8dc3-fcb6a84733b2.png">

We're trying to provide the same experience for free Atomic sites as the one we provide for Simple free. These settings weren't mentioned in the issue, but they're also Marketing -> Traffic settings that shouldn't be available / visible on Atomic free. I think there might be additional issues, like whether we show the ad removal upgrade banner and Cloudflare.

mmtr replied:
Gotcha. As per the discussion in pd2C1Z-10-p2, I think it makes sense to remove/hide these features from Atomic sites on any plan (free or paid) in order to achieve parity across Simple and Atomic sites (although I recommend allowing users to turn on the features if they were deactivated in the past).

As per the helper function used for determining if a site is eligible for a feature, I don't see a clear correlation between WP.me shortlinks or Stats, so I like your suggestion of creating a more abstract `hasSiteAnalyticsFeature` function which checks if a given feature is available for a given site (e.g. `hasSiteAnalyticsFeature( state, siteId, 'shortlinks' )`).

allilevine replied:
@mmtr I've removed this change from this PR in favor of investigating removing these from Atomic further. Follow-up issues:
- https://github.com/Automattic/wp-calypso/issues/53507
- https://github.com/Automattic/wp-calypso/issues/53508

mmtr commented on client/my-sites/site-settings/analytics/form-google-analytics-simple.jsx:
```diff
@@ -73,9 +75,11 @@ const GoogleAnalyticsSimpleForm = ( {
-		const plan = findFirstSimilarPlanKey( site.plan.product_slug, {
-			type: TYPE_PREMIUM,
-		} );
+		const plan = ! isSEOEligible
```
This also confuses me. What's the connection between the SEO and the GA features?

allilevine replied:
This is a check for which plan slug to use in the URL (use `value-bundle` for free Atomic). We could change it to an Analytics eligibility check.

mmtr replied:
Since the plan slug actually depends on the site (`value-bundle` for WP.com sites, `jetpack_security_daily_monthly` for Jetpack sites) and not on the feature, I'd use a site check here. I think that's less confusing.

allilevine replied:
Thanks, I've changed to checking for a WPCOM site with a free plan here. Otherwise `findFirstSimilarPlanKey()` returns a Jetpack plan. 

For all other WPCOM plans we're checking the current plan duration to determine which slug to use.

mmtr commented on client/my-sites/site-settings/analytics/form-google-analytics.jsx:
```diff
@@ -88,8 +90,9 @@ export const GoogleAnalyticsForm = ( props ) => {
+		isSEOEligible,
 	};
-	if ( props.siteIsJetpack ) {
+	if ( props.siteIsJetpack && isSEOEligible ) {
```
Same here, I don't understand why the SEO eligibility is affecting the GA eligibility.

allilevine replied:
This is a check for which Analytics form to show. For Atomic, we show the Jetpack form. But for free Atomic, we want to show the Simple form with WordPress.com plan upgrade nudges. We could change it to an Analytics eligibility check.

mmtr replied:
> We could change it to an Analytics eligibility check.

👍

allilevine replied:
Changed here: https://github.com/Automattic/wp-calypso/pull/53196/commits/1cf6ee4b7abc92aeaa02d91ca239f8992bf10e1a

mmtr reviewed (commented):
Nice! Tested this on a variety of scenarios (Free Simple, Premium Simple, Business Simple, Free Atomic, Business Atomic, Free Jetpack) and they all worked as expected!

Got some questions code-wise though, since I'm really confused about the dependency between the SEO and the Google Analytics features.

allilevine committed:
Return false if missing data for eligibility check

pablinos commented on client/lib/site/utils.js:
```diff
@@ -176,3 +177,19 @@ export function hasSiteFeature( site, feature ) {
+	}
+
+	return ! ( isFreePlan( site.plan.product_slug ) && isSiteWPCOM( state, siteId ) );
+}
```
I was doing something similar in #53216, but when I started needing to use state, I decided that maybe my equivalent helper function should [live with the other selectors](https://github.com/Automattic/wp-calypso/pull/53216/files#diff-a127bedefab489a412871109c45762ff55c5b2e0bfe0ca7f853f8f3bccce253bR1).

Could it be worth doing the same thing here, or is it better placed in this utils lib?

pablinos replied:
Not sure what happened here, this looks like from a review I started on a previous version!

It's the same point I've made below, but it's not as important as that makes it appear - I can see pros and cons for leaving these functions here as well as moving it.

allilevine committed:
Clarify and move utils
Group SEO and Analytics utils in site-settings utils, change eligibility
check to free WPCOM check.

allilevine committed:
Change isSEOEligible to isFreeWPCOM

allilevine committed:
Remove shortlinks & stats changes, use analytics check

mmtr commented on client/my-sites/site-settings/analytics/form-google-analytics.jsx:
```diff
@@ -88,8 +90,9 @@ export const GoogleAnalyticsForm = ( props ) => {
+		isFreeWPCOM,
 	};
-	if ( props.siteIsJetpack ) {
+	if ( props.siteIsJetpack && isGoogleAnalyticsEligible ) {
```
I think `isGoogleAnalyticsEligible` returns `false` for Jetpack Free sites and that causes the regression I noted.

mmtr commented on client/my-sites/site-settings/seo-settings/form.jsx:
```diff
@@ -306,23 +308,25 @@ export class SeoForm extends React.Component {
-						} ),
-			  };
+		const upsellProps =
+			siteIsJetpack && ! isFreeWPCOM
```
Non-blocking: In the future, Atomic sites could support Personal and Premium plans, which are not eligible for the SEO feature. Would it be more robust to check `! isAtomic` here to avoid potential issues?

allilevine replied:
That's a good idea, changed here: https://github.com/Automattic/wp-calypso/pull/53196/commits/69530633390ae28bd73864fc4d486844661f2423

mmtr reviewed (commented):
Tested this on a Simple Free site, Simple Business site, Atomic Free site, Atomic Business site, and Jetpack Free site. They all worked great except the Jetpack Free site, which has this regression:

Before | After
--- | ---
<img width="1239" alt="Screen Shot 2021-06-09 at 11 54 03" src="https://user-images.githubusercontent.com/1233880/121334412-ed727a00-c919-11eb-8366-c40d1cd0100e.png"> | <img width="892" alt="Screen Shot 2021-06-09 at 11 54 34" src="https://user-images.githubusercontent.com/1233880/121334433-f1060100-c919-11eb-8588-5a403c3b75f4.png">

- Copy of the nudge has changed from `Connect your site to Google Analytics` to `Connect your site to Google Analytics in seconds with the Premium plan` .
- URL has changed from `/checkout/:site/jetpack_security_daily_monthly` to `/plans/:site?feature=google-analytics&plan=jetpack_premium`

mmtr commented on client/my-sites/site-settings/utils.ts:
```diff
@@ -13,9 +20,48 @@ type Site = {
+	state: string,
+	siteId: string
+): boolean | undefined {
+	if ( state && siteId && isFreeWPCOMSite( site, state, siteId ) ) {
+		return false;
+	}
```
Trying to understand why is this needed, it seems that this handles an underlying issue where the plan for Atomic Free sites is set incorrectly as `jetpack_free`, right?

<img width="863" alt="Screen Shot 2021-06-09 at 13 10 06" src="https://user-images.githubusercontent.com/1233880/121344716-5bbc3a00-c924-11eb-9bd5-e37caacc2f67.png">

Would it be possible to fix the root issue in WP.com, so the proper plan is returned? Maybe after doing that we could remove most of these workarounds?

```diff
Index: wp-content/admin-plugins/wpcom-billing/class.wpcom-store-api.php
===================================================================
--- wp-content/admin-plugins/wpcom-billing/class.wpcom-store-api.php    (revision 226908)
+++ wp-content/admin-plugins/wpcom-billing/class.wpcom-store-api.php    (working copy)
@@ -443,7 +443,7 @@
                                $billing_period = 'Biyearly';
                        }
                } else {
-                       $product_id = is_jetpack_site( $blog_id ) ? JETPACK_FREE : WPCOM_FREE_PRODUCT;
+                       $product_id = is_jetpack_site( $blog_id ) && ! is_blog_atomic( get_blog_details( $blog_id ) ) ? JETPACK_FREE : WPCOM_FREE_PRODUCT;
                        $product = WPCOM_Store::get_product( $product_id );
                        $subscription_is_expired = false;
                        $user_is_owner = false;
```

allilevine replied:
> Trying to understand why is this needed, it seems that this handles an underlying issue where the plan for Atomic Free sites is set incorrectly as `jetpack_free`, right?

That's correct, and the issue was discussed here: pd2C1Z-Q-p2

> Would it be possible to fix the root issue in WP.com, so the proper plan is returned? Maybe after doing that we could remove most of these workarounds?

That would be awesome! My understanding is that it's a bigger task to make that change and check that it doesn't break anything and that all of the feature checks continue to work. This marketing settings fix is on our list of issues blocking the free Atomic work, so we're looking for a short-term UI workaround.

allilevine committed:
Fix Jetpack free regression, check for Atomic for SEO upsell

allilevine commented:
> Tested this on a Simple Free site, Simple Business site, Atomic Free site, Atomic Business site, and Jetpack Free site. They all worked great except the Jetpack Free site, which has this regression:

Thanks for catching that! I forgot that was why I couldn't use `isGoogleAnalyticsEligible` for which form to show. Fixed here: https://github.com/Automattic/wp-calypso/pull/53196/commits/69530633390ae28bd73864fc4d486844661f2423

pablinos commented on client/my-sites/site-settings/seo-settings/form.jsx:
```diff
@@ -479,15 +485,15 @@ export class SeoForm extends React.Component {
 
 const mapStateToProps = ( state ) => {
-	const selectedSite = getSelectedSite( state );
+	const site = getSelectedSite( state );
```
Is this renamed to make it more concise?

allilevine replied:
Mostly because it was annoying me that it wasn't consistent with the other props. 😅 Should I revert it to keep the changes down?

pablinos replied:
I'm not sure. Consistency is a valid reason to rename it too! It gets renamed to `selectedSite` again later on. Would we want to change that too?

pablinos commented on client/my-sites/site-settings/utils.ts:
```diff
@@ -13,9 +20,48 @@ type Site = {
+ * @param siteId Site ID to check.
+ * @returns bool True if the site is on a free pln and a WPCOM site, false otherwise.
+ */
+export function isFreeWPCOMSite( site: Site, state: string, siteId: string ): boolean | null {
```
Now this is using state, it's essentially a selector. Should we move these helpers to live with the other selectors?

allilevine replied:
I moved / converted them to selectors: https://github.com/Automattic/wp-calypso/pull/53196/commits/0e63e216afb7d078af1935ed40c02364ae1140b7

I'm getting this warning though: "Do not pass complex objects as arguments for a memoized selector". Maybe `site` is too complex. 🤔 

I should probably also add selector tests to replace the util tests I deleted.

pablinos reviewed (commented):
Looks good! Just a couple of queries, but it tests ok and shows right nudges.

I notice the nudge talks about a premium plan, but then that's not available when you click through. I think that's something we'll have to live with for now until we have all plans available on Atomic.

allilevine committed:
Convert util helpers to selectors

allilevine commented:
> I notice the nudge talks about a premium plan, but then that's not available when you click through. I think that's something we'll have to live with for now until we have all plans available on Atomic.

Yeah, I've opened an issue: https://github.com/Automattic/wp-calypso/issues/53515

allilevine committed:
Use simple selectors instead of memoized

allilevine committed:
Add selector tests
- Add tests for new selectors: is-site-wpcom-on-free-plan,
  has-site-analytics-feature, has-site-seo-feature
- Fix feature checks for enterprise and VIP
- Add additional checks for state and siteId to avoid errors

allilevine committed:
Fix selector test names, mock data

allilevine committed:
Remove unused fixtures and add free Atomic test

pablinos commented on client/my-sites/site-settings/analytics/form-google-analytics.jsx:
```diff
@@ -88,8 +90,9 @@ export const GoogleAnalyticsForm = ( props ) => {
+		isFreeWPCOM,
 	};
-	if ( props.siteIsJetpack ) {
+	if ( props.siteIsJetpack && ! isFreeWPCOM ) {
```
I wonder whether in a future change we should adapt the `hasSiteAnalyticsFeature` selector to return something like `wpcom` and `jetpack`, so that we can keep this logic contained in the selector. 

Eventually that logic could also be pushed server-side, extending the plan features work @retrofox has been doing.

allilevine replied:
That would definitely be an improvement. I've also wondered if we should continue to have two different Google Analytics form components.

pablinos commented on client/my-sites/site-settings/seo-settings/form.jsx:
```diff
@@ -306,23 +311,25 @@ export class SeoForm extends React.Component {
+						),
+						feature: FEATURE_ADVANCED_SEO,
+						plan: isFreeWPCOM
+							? PLAN_BUSINESS
```
Is this because Atomic Free sites will be on the `jetpack_free` plan which `findFirstSimilarPlanKey` will fail for? If so, should this check be `isAtomic` to prevent us from changing the code run for Simple Free sites?

allilevine replied:
Yes. Maybe that needs to be clearer.

I added `isAtomic`, so now it checks `isFreeWPCOM && isAtomic`. I think in general on Atomic plans we still want to check what term their plan is for, e.g. 2-year, 1-year, monthly to find the right plan slug.

pablinos commented on client/my-sites/site-settings/seo-settings/form.jsx:
```diff
@@ -306,23 +311,25 @@ export class SeoForm extends React.Component {
-						} ),
-			  };
+		const upsellProps =
+			siteIsJetpack && ! isAtomic
```
Similarly to the analytics checks, it feels like this logic shouldn't be here. We should probably have the plan to promote, current plan, and site type all sent from the server with the other feature information, but that's definitely not something to tackle in this PR.

Maybe we should decide if that's a good idea and raise an issue.

allilevine replied:
I opened an issue to track it: https://github.com/Automattic/wp-calypso/issues/53793

pablinos commented on client/my-sites/site-settings/seo-settings/form.jsx:
```diff
@@ -479,15 +486,15 @@ export class SeoForm extends React.Component {
 	const isAdvancedSeoEligible =
-		selectedSite.plan &&
-		hasSiteSeoFeature( selectedSite ) &&
+		site.plan &&
```
Maybe it's good to short circuit the check early, have we implicitly got this check in `hasSiteSeoFeature`?

allilevine replied:
Yeah I think we do, I removed that check here: https://github.com/Automattic/wp-calypso/pull/53196/commits/6c722ae335fe3176cbfe6cd5c57698efe9b0f536

I also reverted the `selectedSite` -> `site` change since we're no longer using `site` with the selectors.

pablinos reviewed (commented):
This is looking great, thanks for doing all that refactoring!

Just a few minor comments and thoughts, but this is pretty much there I reckon.

allilevine committed:
Clean up seo-settings form props and checks

allilevine committed:
Use site.plan instead of getCurrentPlan, to fix VIP

allilevine committed:
Fix tests and add Atomic check for correct plan

allilevine commented on client/state/selectors/has-site-analytics-feature.js:
```diff
@@ -0,0 +1,30 @@
+ * @param {number} siteId The id of the site to check.
+ * @returns {boolean} True if the site has the analytics feature, false otherwise.
+ */
+export default ( state, site, siteId ) => {
```
This is messy but I wanted to use the data we already have, rather than re-creating the `site` object in the selector. Let me know if you have any suggestions for improving it.

pablinos replied:
I know we've talked a little about this, but I've just been doing a bit more thinking, and I'm leaning on sticking to using just the `siteId`.

From what I can see of the other selectors we're only ever using the site's ID, so this would make it consistent and I notice that we're using `getSelectedSite` only to pass it to these selectors, so we could get the site object when we need it and default the site ID parameter to the currently selected site - `siteId = getSelectedSiteId( state )` - which seems to be a common pattern.

allilevine replied:
Refactored to use just `siteId` here: https://github.com/Automattic/wp-calypso/pull/53196/commits/4d964cf3942b6bde035d85093b19ee22403dbb6f

I had to the remove tests for no data passed in since `getSelectedSiteId( state )` requires `state`.

retrofox reviewed (commented):
LGTM. I've tested the approach and it works as expected. Also, we talked about a possible follow-up task where we may want to make focus on the site features instead of the site type and plan but it's something that we can iterate later.

Thanks for taking over this tough issue. 👏

retrofox reviewed (approved):
Forgot to approve it. ✅

allilevine committed:
Refactor to pass in siteId, not site

allilevine committed:
Rebase layout/index.js

allilevine committed:
Remove security selector file

pablinos commented on client/layout/index.jsx:
```diff
@@ -295,6 +296,7 @@ class Layout extends Component {
 				<QuerySites primaryAndRecent={ ! config.isEnabled( 'jetpack-cloud' ) } />
 				{ this.props.shouldQueryAllSites && <QuerySites allSites /> }
 				<QueryPreferences />
+				<QuerySiteFeatures siteId={ this.props.siteId } />
```
I'm not sure where it's triggered (perhaps as part of QuerySites) but the features information is being returned under the `plan` section of the data that comes back from the `/sites/<SITE ID>` API call. Could we save a request by using that data and only explicitly querying for it if it's unavailable, or we suspect it's stale?

allilevine replied:
I decided to try this out by checking for the features under the site plan data in the selector, and then checking that selector before rendering the `<QuerySiteFeatures>` component. If we already have the features from either /sites or /features, then we shouldn't fetch the features. I also prevented fetching if we were already fetching /site or /sites.

What kept happening was that state would update with the features from /sites, but then update without them afterwards. And then of course we'd have to fetch the features again.

The cause seems to be that while the `v1.1/sites/<SITE ID>` API response does include features in the plan section, the `v1.2/me/sites` API response we fetch afterwards on line 297 (`<QuerySites allSites />`) does not include features. That replaces the site data so that features are removed from state. 

Should we update the `v1.2/me/sites` endpoint to include features? Or is there another way of working around this that I missed?

retrofox replied:
Also, for some reason sometimes, the `plan` section is not populated in the endpoint response. I tested with retrosimplepremium01.wordpress.com site.

allilevine committed:
Remove get-available-plan-upgrade

allilevine committed:
Add tests for seo-settings/help and cloudflare analytics

retrofox commented on client/my-sites/domains/domain-management/list/index.jsx:
```diff
@@ -472,7 +471,6 @@ export class List extends React.Component {
 
 		return [
 			<QuerySitePurchases key="query-purchases" siteId={ selectedSite.ID } />,
-			<QuerySiteFeatures key="query-features" siteId={ selectedSite.ID } />,
```
nice

retrofox reviewed (approved):
Tested again, and it LGTM.
Confirmed that the app hits the features endpoint properly. Also, confirmed the different cases described in the PR description  👍
